docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,198 @@
+* `-` If an existing app no longer uses KBP, their app is still limited unless the owner lets us know.
+* `-` Nodes will continue be taken up by jobs that do little computation as each job gets its own node.
+* `-` Users may not be able to effectively use up their 10 job spots 
+* `-` Capacity is still wasted, as some KBP jobs sit around waiting for other jobs to run
+
+###  LIMIT KBP jobs to a maximum of N<10 active jobs per user + Seperate queue for kbparallels apps 
+* `+` Same pros as above
+* `+` Allows us to group up KBP jobs onto fewer machines, instead of giving them their entire node
+* `-` Requires going through each app and understanding the worst case computational needs in order to  set the estimated cpu and memory needs for each app 
+* `-` Apps can interfere with other innocent apps and take them down
+* `-` Creating a new queue requires balancing between how many active KBP nodes there vs how many nodes are available for other NJS jobs.
+* `-` Capacity is still wasted, as some KBP jobs sit around waiting for other jobs to run
+
+
+### Build KBP Lightweight Version + KBP Queue
+
+
+#### Design of new verison
+* All apps must be modified to use the new KBP lightweight version, which will: 
+* Can either modify KBP, or create a new tool/package to use instead of KBP
+
+1) Launch a management job called the *Job Manager* that sits in the KBP Queue, alongside other KBP jobs. Other jobs are launched in the NJS queue.
+2) Launch the *Setup Job* which will
+ * Use the *User Parameters* and/or
+ * Optionally Download the Data from the *User Parameters*  to figure out *Job Manager* parameters
+ * Use the results of information gathered from the initial download and or *User Parameters*
+ * Generate final parameters to be sent to the *Job Manager* to launch *Fan Out* jobs
+3) The *Job Manager* now has enough parameters to setup *Fan Out* Jobs, and will send out jobs and wait for them (and possibly retry them upon failure)
+4) *Fan Out* jobs download data and perform calculations, save them back to the system, and return references to the saved objects
+5) The *Job Manager* optionally launches a *Group/Reduce* job based on User Parameters and or the results of *Fan Out* Jobs
+6) The *Group/Reduce* job downloads objects from the system, and creates a set or other grouping, and then saves the set back to the system
+7) The *Job Manager*  sends the results back to a *Report* Job, which downloads the final data and uploads a report based on that data
+8) The *Job Manager* returns the reference to the results of the *Report Job*
+
+Pros/Cons
+* `-` Addresses the deadlocking issue, UI still broken for regular runs and batch runs if we re-use KBP
+* `-` Have to rewrite all apps that use KBP to use this new paradigm
+* `-` Will have to upload data and download data more times as no computation is done in the *Job Manager*
```

Comment:
I'm not sure I understand why this is is true. In current KBP, the data transfers are:
1) KBP node downloads data to calculate jobs - this will happen the same amount of times in LKBP (lightweight KBP), but in the setup job vs. the KBP job
2) PIP nodes download data, same as LKBP
3) KBP node downloads data to summarize / report - same for LKBP in the group/reduce if that makes the report as well

Reply:
>  Will have to upload data and download data more times as no computation is done in the *Job Manager*
If the group reduce is done on objects, then no files have to be downloaded
if the group reduce is done on files that need to be downloaded from objects and then a new object is created you have to download these files to disk first.

Reply:
But that's true for both KBP and LKBP, unless I'm misunderstanding

Reply:
Sounds like we should do another call

Reply:
:+1: